src/complex/Utilities/OStreamUtilities.hpp

```diff
@@ -0,0 +1,765 @@
+  }
+};
+
+/**
+ * @brief Parallel enabled | implicit loading of **DataArray**'s elements to PrintMatrix2D vertically
+ * @tparam T The primitive type attacthed to **DataArray**
+ * @param outputMatrix The PrintMatrix2D that will store string values from *inputArray*
+ * @param inputArray The **DataArray** that will have its values translated into strings
+ * @param columnToWrite The column that the **DataArray** should be stored in
+ * @param hasHeaders bool to determine if index must be advanced 1 to avoid overwrite
+ */
+template <typename T>
+struct LoadDataArrayToMatrixImpl
+{
+  LoadDataArrayToMatrixImpl(PrintMatrix2D& outputMatrix, const DataArray<T>& inputArray, const size_t& columnToWrite, bool hasHeaders = false)
+  : m_OutputMatrix(outputMatrix)
+  , m_ColumnToWrite(columnToWrite)
+  , m_InputArray(inputArray)
+  , m_HasHeaders(hasHeaders)
+  {
+  }
+  ~LoadDataArrayToMatrixImpl() = default;
+
+  void loadToMatrix(size_t start, size_t end) const
+  {
+    if(m_HasHeaders)
+    {
+      size_t row;
+      for(size_t i = start; i < end; i++)
+      {
+        row = i + 1;
+        m_OutputMatrix.getValue(row, m_ColumnToWrite) = StringUtilities::number(m_InputArray[i]);
+      }
+    }
+    else
+    {
+      for(size_t i = start; i < end; i++)
+      {
+        m_OutputMatrix.getValue(i, m_ColumnToWrite) = StringUtilities::number(m_InputArray[i]);
+      }
+    }
+  }
+  void operator()(const Range& range) const
+  {
+    loadToMatrix(range.min(), range.max());
+  }
+
+private:
+  PrintMatrix2D& m_OutputMatrix;
+  const size_t& m_ColumnToWrite;
+  const DataArray<T>& m_InputArray;
+  bool m_HasHeaders = false;
+};
+
+/**
+ * @brief implicit loading of **NeighborList**'s elements to PrintMatrix2D
+ * @tparam T The primitive type attacthed to **NeighborList**
+ * @param outputMatrix The PrintMatrix2D that will store string values from *inputArray*
+ * @param inputArray The **NeighborList** that will have its values translated into strings
+ * @param hasIndex bool to determine if index must be advanced 1 to avoid row overwrite
+ * @param hasHeaders bool to determine if index must be advanced 1 to avoid column overwrite
+ */
+template <typename T>
+struct LoadNeighborListToMatrixImpl // adds component count implicitly
+{
+  LoadNeighborListToMatrixImpl(PrintMatrix2D& outputMatrix, const NeighborList<T>& inputArray, bool hasIndex = false, bool hasHeader = false)
+  : m_OutputMatrix(outputMatrix)
+  , m_InputArray(inputArray)
+  , m_HasIndex(hasIndex)
+  , m_HasHeader(hasHeader)
+  {
+  }
+  ~LoadNeighborListToMatrixImpl() = default;
+
+  void loadToMatrix(size_t start, size_t end) const // use initial logic checking to avoid repeating ifs in loops
+  {
+    size_t row, column;
+    if(m_HasHeader)
+    {
+      if(m_HasIndex) // yes index yes header
+      {
+        for(size_t i = start; i < m_InputArray.getNumberOfLists(); i++)
+        {
+          row = i + 1;
+          auto grain = m_InputArray.getListReference(i);
```

Comment:
```suggestion
          const auto& grain = m_InputArray.getListReference(i);
```
In this case, `auto` would be a plain `std::vector<T>` which would make a copy. In general it is important to specify the const-ness and whether it is a pointer or reference when using `auto` to avoid accidental copies.